aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1689,5 +1690,89 @@ Tensor chain_matmul(TensorList matrices) {
   }
 }
 
+DEFINE_DISPATCH(unpack_pivots_stub);
+
+std::tuple<Tensor, Tensor, Tensor> _lu_unpack(
+    const Tensor& LU_data,
+    const Tensor& LU_pivots,
+    bool pivots_as_permutation_matrix
+    ) {
+  // In the generalized LU factorization, the following shape relations hold:
+  // A.shape[-2:] == (m, n),
+  // P.shape[-2:] == (m, m),
+  // U.shape[-2:] == (m, k),
+  // L.shape[-2:] == (k, n),
+  // where k = min(m, n)
+  int64_t m = LU_data.size(-2);
+  int64_t n = LU_data.size(-1);
+  int64_t k = std::min(m, n);
+
+  auto U = LU_data.triu();
+  if (m != k) {
+    U = U.narrow(-2, 0, k);
+  }
+
+  auto L = LU_data.tril();
+  if (k != n) {
+    L = L.narrow(-1, 0, k);
+  }
+  L.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(1);
+
+  auto unpacked_pivots_sizes = LU_pivots.sizes().vec();
+  unpacked_pivots_sizes[LU_pivots.dim() - 1] = m;
+  auto unpacked_pivots = at::zeros(
+    unpacked_pivots_sizes,
+    LU_pivots.options().memory_format(at::MemoryFormat::Contiguous)
+  );
+
+  // Fill `unpacked_pivots` with identity permutation
+  auto id_perm = at::arange(m, LU_pivots.options());
```

Comment:
Why not create `unpacked_pivots` directly with arange (+ a .view() if needed)?

Reply:
You need to sequentially swap elements from left to right.

Reply:
Ho I don't mean that final thing, I mean why create it above and then copy the arange result into it here. that's it.

Reply:
I need to broadcast the identity permutation to all matrices in a batch, because each matrix has its own `pivots[batch_idx]`.

Reply:
Ho there is hidden broadcasting in the copy that makes it convenient. Ok.